ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -7,62 +7,77 @@ module Ouroboros.Network.Mux (
+    , VersionNumber (..)
     , RemoteClockModel (..)
     , encodeMuxSDU
     , decodeMuxSDUHeader
+    , muxBearerSetState
     , muxStart
     ) where
 
+import qualified Codec.CBOR.Read as CBOR
+import           Codec.CBOR.Write (toLazyByteString)
 import           Control.Monad
 import           Control.Monad.Class.MonadAsync
 import           Control.Monad.Class.MonadFork
 import           Control.Monad.Class.MonadSay
 import           Control.Monad.Class.MonadSTM
 import           Control.Monad.Class.MonadThrow
-import           Control.Monad.Class.MonadTimer
 import           Data.Array
 import qualified Data.ByteString.Lazy as BL
-import           Data.Word
+import           Data.List (intersect)
+import qualified Data.Map.Strict as M
+import           Data.Maybe (fromJust)
+import           GHC.Stack
 
 import           Ouroboros.Network.Channel
-
+import           Ouroboros.Network.Mux.Control
 import           Ouroboros.Network.Mux.Egress
 import           Ouroboros.Network.Mux.Ingress
 import           Ouroboros.Network.Mux.Types
 
--- | muxStart starts a mux bearer for the specified protocols with the provided read and write
--- functions.
+
+-- | muxStart starts a mux bearer for the specified protocols corresponding to
+-- one of the provided Versions.
 -- TODO: replace MonadSay with iohk-monitoring-framework.
-muxStart :: (MonadAsync m, MonadFork m, MonadSay m, MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
-        => MiniProtocolDescriptions ptcl m
-        -> (MuxSDU ptcl -> m (Time m))          -- Write function
-        -> m (MuxSDU ptcl, Time m)              -- Read function
-        -> m Word16                             -- SDU size function
-        -> m ()                                 -- Close function
+muxStart :: ( MonadAsync m, MonadFork m, MonadSay m, MonadSTM m, MonadThrow m
+            , Ord ptcl, Enum ptcl, Bounded ptcl)
+        => [(Version, MiniProtocolDescriptions ptcl m)]
+        -> MuxBearer ptcl m
+        -> MuxStyle
         -> Maybe (Maybe SomeException -> m ())  -- Optional callback for result
-        -> m (ThreadId m)
-muxStart udesc wfn rfn sdufn close  rescb_m = do
+        -> m ()
+muxStart verMpds bearer style rescb_m = do
     tbl <- setupTbl
     tq <- atomically $ newTBQueue 100
     cnt <- newTVarM 0
-    let pmss = PerMuxSS tbl tq wfn rfn sdufn
+    (_, udesc) <- if style == StyleClient then muxClient verMpds bearer
+                                          else muxServer verMpds bearer
+
+    let pmss = PerMuxSS tbl tq bearer
         jobs = [ demux pmss
                , mux cnt pmss
                , muxControl pmss ModeResponder
                , muxControl pmss ModeInitiator
                ]
-    mjobs <- sequence [ mpsJob cnt pmss (udesc ptcl)
+    mjobs <- sequence [ mpsJob cnt pmss udesc ptcl
                       | ptcl <- [minBound..maxBound] ]
     aids <- mapM async $ jobs ++ concat mjobs
-    fork $ watcher aids
+    muxBearerSetState bearer Mature
+    watcher aids
```

Comment:
I think this means that the `muxStart` will block now while watching the other threads.  It would be good to mention this in the documentation that the call will block until all the running mux jobs will terminate (or one of them will raise an exception) ...

Reply:
It will block until _any_ mux job terminates not all. Will add a comment.

Reply:
Yes, you're right, I wasn't precise.